### PR TITLE
Targeted Grant: Native Protocol Handler API for Browser Extensions

### PR DESCRIPTION
**Pull request by lidel**
> ### Project Description
> 
> Browser extensions are unable to register handlers for URIs such as `ipfs://<cid>`  
> 
> This grant aims to create a new API that enables browser extension to register
> a _native_ protocol handler capable of returning streaming responses without redirecting to a third party server.

This is wip, needs input from Igalia
cc @autonome 

PREVIEW: https://github.com/ipfs/devgrants/blob/protocol-handler-api-for-browser-extensions/targeted-grants/protocol-handler-api-for-browser-extensions.md

### CODE REVIEW

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+```
+
+Above example:
+
+- Works only when executed on matching Origin
+    - User needs to be on `example.com` to registration to work
+    - Displays user prompt, asking for confirmation if a handler should be registered
+- Requires unknown protocols names to be prefixed with `web+` ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler#Permitted_schemes))
+   - Firefox whitelisted `ipfs://` ([bug 1428446](https://bugzilla.mozilla.org/show_bug.cgi?id=1428446))
+   - Chromium requires prefix `web+ipfs://` ([bug 651311](https://bugs.chromium.org/p/chromium/issues/detail?id=651311), [intent-to-implement](https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/29sFh4tTdc))
```
Comment (fred-wang) [inline targeted-grants/protocol-handler-api-for-browser-extensions.md]: The link to the intent-to-implement seems broken. Are you able to provide the correct link or at least the feature name?

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+| 2 | PoC API implementation in Chromium | $X | Y weeks |
+| 3 | Reusable Chromium/Blink patches work with Brave | $X | Y weeks |
+| 4 | ? | $X | Y weeks |
+| 5 | ?  | $X | Y weeks |
+
+## Acceptance Criteria
+
+<!-- What are the acceptance criteria for each milestone and for the final deliverables? These should be as objective as possible. They will be used to determine whether or not a grantee will receive payment for work completed for a milestone.  -->
+
+- API specificiation document is approved by IPFS project
```
Comment (fred-wang) [inline targeted-grants/protocol-handler-api-for-browser-extensions.md]: nit: specification

Do we want to standardize this specification? Is there a place where to put WebExtensions API specifications (where does Mozilla put such specifications) ?

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    - Origin is based on the content root
+    - supports streaming responses by means of async iterators
+    - reuses HTTP semantics for caching, content type, headers and error codes
+- API implementation in form of patches for Chromium codebase
+    - allows JS running in browser extension context to register `ipfs://` and `ipns://` protocol handlers, process every request made with them and return arbitrary bytes 
+    - released under [PL's Permissive License Stack](https://protocol.ai/blog/announcing-the-permissive-license-stack/)
```
Comment (fred-wang) [inline targeted-grants/protocol-handler-api-for-browser-extensions.md]: I would need to read more about this license, but any code that is intended to land in Chromium (or WebKit and Gecko) should be under an open source license and more specifically the license (and sometimes copyright holder) suggested by the corresponding project. For Chromium it's the BSD-like one at https://chromium.googlesource.com/chromium/src.git/+/refs/heads/master/LICENSE and copyrighted "The Chromium Authors".

I guess the text of the license might also need to be a under some open license, maybe Mozilla has rules too.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+The work in scope of this grant is to leverage lessons from the past experiments
+and create a general purpose Protocol Handler API that can be used by any browser extension.
+Majority of browser vendors support a variant of Chromium's [Manifest V2](https://developer.chrome.com/extensions/manifest)
+browser extensions (Firefox calls them [WebExtensions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions)).
+Chromium-based vendors are vocal about being interested in improving IPFS support. 
+To maximize the value created by this grant Chromium codebase should be used as implementation target.
+
+If possible, the API should be compatible with `manifest.json/protocol_handlers` API already present in Firefox 
+(registering handler on extension install) but with option to omit `uriTemplate` and provide a self-hosted,
+programmatic handler via `chrome.protocol.registerProtocol` instead.
```
Comment (fred-wang) [inline targeted-grants/protocol-handler-api-for-browser-extensions.md]: Do you have a reference link for `chrome.protocol.registerProtocol`? It does not seem obvious to me what this paragraph is referring to.

**Comment by fred-wang** [comment]
cc @bkardell

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+navigator.registerProtocolHandler('web+ipfs',
+                                  'https://example.com/?uri=%s',
+                                  'IPFS handler')
+```
+
+Above example:
+
+- Works only when executed on matching Origin
+    - User needs to be on `example.com` to registration to work
+    - Displays user prompt, asking for confirmation if a handler should be registered
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Install process displays the list of permissions requested by the extension, for example on which URLs it is allowed to run etc. I imagine on the UX side provided protocol schemes could  be just another item on that list (eg. `ipfs://*`).

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+## Acceptance Criteria
+
+<!-- What are the acceptance criteria for each milestone and for the final deliverables? These should be as objective as possible. They will be used to determine whether or not a grantee will receive payment for work completed for a milestone.  -->
+
+- API specificiation document is approved by IPFS project
+    - Origin is based on the content root
+    - supports streaming responses by means of async iterators
+    - reuses HTTP semantics for caching, content type, headers and error codes
+- API implementation in form of patches for Chromium codebase
+    - allows JS running in browser extension context to register `ipfs://` and `ipns://` protocol handlers, process every request made with them and return arbitrary bytes
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: It was mentioned in https://github.com/whatwg/html/issues/3935, but yes, we usually just use `ipfs://` as its much easier to explain, reason about and test (it is accepting well-defined [CID](https://github.com/ipld/cid) identifiers and data behind each address is immutable)

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+- API specificiation document is approved by IPFS project
+    - Origin is based on the content root
+    - supports streaming responses by means of async iterators
+    - reuses HTTP semantics for caching, content type, headers and error codes
+- API implementation in form of patches for Chromium codebase
+    - allows JS running in browser extension context to register `ipfs://` and `ipns://` protocol handlers, process every request made with them and return arbitrary bytes 
+    - released under [PL's Permissive License Stack](https://protocol.ai/blog/announcing-the-permissive-license-stack/)
+- API proliferation
+    - can be enabled at a build time by browser vendors such as Brave or Edge
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Putting it behind a runtime flag for upstream Chromium is fine. 
We just want to have an easy way of enabling the feature by default if a vendor is ok with shipping it before Chromium does.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    - patches submitted to the Brave project
+    - patches submitted to the upstream Chromium / Blink projects
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Yeah, it was unclear to me if browser extension APIs fall under "web API":

> You do not need to use this process if:
> - Your change does not affect web API behavior to the point that developers need to be aware of it.
> [..]

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 4 above)
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: I meant future API which does not exist yet. Updated the section to make it clear.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 3 above)
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Ack. By default, PL releases everything under both MIT+Apache 2.0 to ease the adoption.  This grant includes contribution to a third party project – we can be flexible here. I've updated this point to allow for a license suggested by Chromium project.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 2 above)
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Yes, we definitely want this API to be created with Mozilla and Chromium being in the loop.
I've added a note to "API proliferation" section.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 1 above)
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Correct. I'll update the groups link and added https://github.com/whatwg/html/issues/3998 to this section. 

We are aware of the stalemate and hope that going browser extension route, discussing a new, generic API could provide incentive to unblock this discussion on the vendor side.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                                  'IPFS handler')
+```
+
+Above example:
+
+- Works only when executed on matching Origin
+    - User needs to be on `example.com` to registration to work
+    - Displays user prompt, asking for confirmation if a handler should be registered
+- Requires unknown protocols names to be prefixed with `web+` ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler#Permitted_schemes))
+   - Firefox whitelisted `ipfs://` ([bug 1428446](https://bugzilla.mozilla.org/show_bug.cgi?id=1428446))
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: I'd keep it here as-is because this whitelisting changed behavior of `navigator.registerProtocolHandler` months before they shipped `manifest.json/protocol_handlers` (which is just a thin orchestration wrapper on top of `navigator.registerProtocolHandler` to make registration work in browser extension context)

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+When defined, [`manifest.json/protocol_handlers`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/protocol_handlers)
+automatically registers a redirect-based protocol handler when browser extension is installed. 
+
+It removes the need for user to visit specific website, 
+but remains a thin wrapper on top of `navigator.registerProtocolHandler` API, 
+with the rest of its limitations: it is impossible for a browser extension
+to return a byte array representing response for requested URL.
+
+This API is not supported by Chromium ([bug 883274](https://bugs.chromium.org/p/chromium/issues/detail?id=883274)).
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Added note on bug 1428446.

Yes, I believe fixing bug 883274 would be a good starting point and definitely could be included in this grant as one of distinct Milestones.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+```js
+navigator.registerProtocolHandler('web+ipfs',
+                                  'https://example.com/?uri=%s',
+                                  'IPFS handler')
+```
+
+Above example:
+
+- Works only when executed on matching Origin
+    - User needs to be on `example.com` to registration to work
```
[the remark this answers is not in the document]
Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: I meant that browser will refuse to register redirect-based handler if the target of the redirect is on a different Origin than JS that called `navigator.registerProtocolHandler`.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (autonome) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: This likely will depend on feedback from Chromium reviewers. We thought a compile-time flag would make it easier to land - means no Chromium-based browser has to ship those bits at all.

A runtime flag is better from our perspective, as it lowers the barrier for experimentation by any given Chromium-based browser. Two of the browsers we're working with have specifically asked for this option.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (autonome) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Is there any precedent for an extension API landing in Chromium but only shipping in non-Chrome browser? Might be entering some uncharted water. Hopefully can make it clearer for those following.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Right, i was not sure about the wording of "to registration to work" but I'm not a native speaker

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Yes, that would be my guess to.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Compile-time flag is generally an issue for upstream work as this can cause build issues that are not found by the official CI-bots. Even when the work is done downstream, I think it would still be preferable to do a runtime flag. The only advantage of compile-time is if we end up adding more third-party dependencies. Anyway that's just a detail, I would suggest rewording "build or run time" to make things more flexible.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: In general, I don't believe any new web platform feature are landing in Chromium without being enabled in Chrome. But again, I'm not sure for web extension, I would need to talk to people who are more familiar with that.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: OK sounds good. I believe re-opening discussion on the WHATWG/Chromium/Mozilla whitelisting / blacklisting for navigator.registerProtocolHandler, could also be a good starting point and a distinct/parallel task.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 4 above)
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: I see. I think this is my main interrogation right now as without further investigation it's a bit difficult to estimate how much work is needed to implement the backend code or if there is anything we can rely on.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: mmh, that's interesting. So are you saying navigator.registerProtocolHandler accepts `ipfs://` and `ipns://` outside Web Extensions? IIRC believed I failed to make it work last week. This would mean Firefox decided to depart from the current spec (which is bad...) and that they forgot to update https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler#Permitted_schemes

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 4 above)
Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: @lidel To clarify my interrogation, https://blog.ipfs.io/2019-10-08-ipfs-browsers-update mentions 

> Chromium OS has a raw sockets API, but it’s disabled in the Chrome browser. Brave has modified their build to whitelist those APIs for IPFS Companion to use - and this gives us superpowers compared to any other browser at this point."

So do you expect we would essentially just enable & expose these ChromiumOS APIs via a new `chrome.*.registerProtocol`? Or is a more significant effort needed on the backend code?

Comment (lidel) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: @fred-wang I know this part is bit confusing, so let me elaborate: `chrome.*` namespace in Chromium-based browsers is populated by both Browser Extension APIs and Chrome OS/Apps APIs, but in this grant we care only about the browser ones.

Situation looks like this:

- chrome apps: https://developer.chrome.com/apps/api_index 
  Avoid in the long term, as Google [deprecated them](https://blog.chromium.org/2020/01/moving-forward-from-chrome-apps.html) and will remove support by 2022.
- browser extensions: https://developer.chrome.com/extensions/api_index
 These APIs were reimplemented by Firefox when they move away from XUL-based extensions.  **This is where we want to add the new protocol handler API**
 We would not change any of _existing_ APIs on `chrome.*`, we would only add a new API to that namespace because that is where other browser extension APIs are)

The new API should enable browser extension to return arbitrary bytes as stream, and if possible, with mandatory/optional content-type (so if content type is stored in IPFS we can avoid mime-type sniffing).

Examples of streaming "protocol handler API":
- https://github.com/mozilla/libdweb#protocol-api
- https://www.electronjs.org/docs/api/protocol#protocol 

We did not write any specifics about the API as we don't want to be overly prescriptive, but we can add something as a starting point, if you feel it will make it easier to discuss this further.

Comment (fred-wang) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md]: Thanks for the clarification ! That's how I had understood the Chrome API stuff, my doubt was about the backend effort/size needed for a "generic Protocol Handler API" exactly. I don't think we need the detailed API in the grant description but probably a simple sketch/example of the desired behavior like you commented here is enough, maybe saying it would be similar to the one mentioned in "Firefox Nightly + libdweb" (incidentally I suspect using modern async iterators is probably what we want). From what I get now, we don't need to add complex network stuff in the backend, we can rely on existing JS async generators or similar and the API would be gluing things together ; most of the actual network work will done by the users.

One thing that confused me was that I forgot this is "generic" and not specific to ipfs at all. Perhaps "Project Description" and "Value" sections should be tweaked a bit to highlight that it would allow people to register *any* (whitelisted) "foo://" protocol, and IPFS ones in particular. I think this is a strong argument to use, for the grant request and for Chromium/Mozilla review.

I'll continue discussion with my colleagues at Igalia and we are going to follow-up privately.

File: targeted-grants/protocol-handler-api-for-browser-extensions.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (parkan) [reply targeted-grants/protocol-handler-api-for-browser-extensions.md:91]: I think "for registration to work" also :)
